File: Stack/232. Implement Queue using Stacks/Solution.py
# N : Total number of elements added
# Time Complexity : Creating MyQueue Object - O(1), push() - O(1), pop() - O(N), peek() - O(N), empty() - O(1)
# Space Complexity : O(N)
# class MyQueue:
#     def __init__(self):
#         self.mainStack = []
#         self.storeStack = []

#     def push(self, x: int) -> None: 
#         self.mainStack.append(x)

#     def pop(self) -> int:
#         n = len(self.mainStack)
#         deletedVal = None
#         for i in range(n):
#             deletedVal = self.mainStack.pop()
#             if i == n - 1:
#                 continue
#             self.storeStack.append(deletedVal)
        
#         while len(self.storeStack) > 0:
#             self.mainStack.append(self.storeStack.pop())

#         return deletedVal

#     def peek(self) -> int:
#         n = len(self.mainStack)
#         deletedVal = None
#         for i in range(n):
#             deletedVal = self.mainStack.pop()
#             self.storeStack.append(deletedVal)
        
#         while len(self.storeStack) > 0:
#             self.mainStack.append(self.storeStack.pop())
            
#         return deletedVal

#     def empty(self) -> bool:
#         return len(self.mainStack) == 0


# N : Total number of elements added
# Time Complexity : Creating MyQueue Object - O(1), push() - O(1), pop() - O(N), peek() - O(N), empty() - O(1)
# Space Complexity : O(N)
class MyQueue:
    def __init__(self):
        self.pushStack = []
        self.popStack = []

    def push(self, x: int) -> None: 
        self.pushStack.append(x)

    def pop(self) -> int:
        n = len(self.pushStack)
        m = len(self.popStack)
        if m == 0:
            for i in range(n):
                self.popStack.append(self.pushStack.pop())
        return self.popStack.pop()

    def peek(self) -> int:
        n = len(self.pushStack)
        m = len(self.popStack)
        if m == 0:
            for i in range(n):
                self.popStack.append(self.pushStack.pop())
        return self.popStack[-1]

    def empty(self) -> bool:
        return len(self.pushStack) == 0 and len(self.popStack) == 0

File: Stack/232. Implement Queue using Stacks/Solution.java
import java.util.*;

// N : Total number of elements added
// Time Complexity : Creating MyQueue Object - O(1), push() - O(1), pop() - O(N), peek() - O(N), empty() - O(1)
// Space Complexity : O(N)
// class MyQueue {
//     Stack<Integer> mainStack;
//     Stack<Integer> storeStack;
//     public MyQueue() {
//         mainStack = new Stack<>();
//         storeStack = new Stack<>();
//     }
    
//     public void push(int x) {
//         mainStack.push(x);
//     }
    
//     public int pop() {
//         int n = mainStack.size();
//         int deletedVal = -1;
//         for (int i=0; i<n; i++){
//             deletedVal = mainStack.pop();
//             if (i == n - 1) continue;
//             storeStack.push(deletedVal);
//         }

//         while (!storeStack.isEmpty()) mainStack.push(storeStack.pop());
//         return deletedVal;
//     }
    
//     public int peek() {
//         int n = mainStack.size();
//         int deletedVal = -1;
//         for (int i=0; i<n; i++){
//             deletedVal = mainStack.pop();
//             storeStack.push(deletedVal);
//         }

//         while (!storeStack.isEmpty()) mainStack.push(storeStack.pop());
//         return deletedVal;
//     }
    
//     public boolean empty() {
//         return mainStack.isEmpty();
//     }
// }


// N : Total number of elements added
// Time Complexity : Creating MyQueue Object - O(1), push() - O(1), pop() - O(1), peek() - O(1), empty() - O(1)
// Space Complexity : O(N)
class MyQueue {
    Stack<Integer> pushStack;
    Stack<Integer> popStack;
    public MyQueue() {
        pushStack = new Stack<>();
        popStack = new Stack<>();
    }
    
    public void push(int x) {
        pushStack.push(x);
    }
    
    public int pop() {
        int m = popStack.size();
        if (m == 0){
            while (!pushStack.isEmpty()) popStack.push(pushStack.pop());
        }
        return popStack.pop();
    }
    
    public int peek() {
        int m = popStack.size();
        if (m == 0){
            while (!pushStack.isEmpty()) popStack.push(pushStack.pop());
        }
        return popStack.peek();
    }
    
    public boolean empty() {
        return pushStack.isEmpty() && popStack.isEmpty();
    }
}
